packages/block-editor/src/components/dimensions-tool/index.js

```diff
@@ -0,0 +1,182 @@
+ * @property {number} [height]      CSS height property.
+ * @property {string} [scale]       CSS object-fit property.
+ * @property {string} [aspectRatio] CSS aspect-ratio property.
+ */
+
+/**
+ * @callback DimensionsControlsOnChange
+ * @param {Dimensions} nextValue
+ * @return {void}
+ */
+
+/**
+ * @typedef {Object} DimensionsControlsProps
+ * @property {string}                     [panelId]            ID of the panel that contains the controls.
+ * @property {Dimensions}                 [value]              Current dimensions values.
+ * @property {DimensionsControlsOnChange} [onChange]           Callback to update the dimensions values.
+ * @property {SelectControlProps[]}       [aspectRatioOptions] Aspect ratio options.
+ * @property {SelectControlProps[]}       [scaleOptions]       Scale options.
+ * @property {WPUnitControlUnit[]}        [unitsOptions]       Units options.
+ */
+
+/**
+ * Component that renders controls to edit the dimensions of an image or container.
+ *
+ * @param {DimensionsControlsProps} props The component props.
+ *
+ * @return {WPElement} The dimensions controls.
+ */
+function DimensionsControls( {
+	panelId,
+	value = {},
+	onChange = () => {},
+	aspectRatioOptions, // Default value is in AspectRatioTool.
+	scaleOptions, // Default value is in ScaleTool.
+	unitsOptions,
+} ) {
+	// TODO: 'spacing.units` was used in the featured image code, but we're
+	// setting width and height, not spacing here, so I'm not sure that's the
+	// right thing to use.
+	const customUnits = useCustomUnits( {
+		availableUnits: useSetting( 'spacing.units' ) || [
+			'px',
+			'%',
+			'vw',
+			'em',
+			'rem',
+		],
+	} );
+	const units = unitsOptions ?? customUnits;
+
+	// 'custom' is not a valid value for CSS aspect-ratio, but it is used in the
+	// dropdown to indicate that setting both the width and height is the same
+	// as a custom aspect ratio.
+	const aspectRatioValue =
+		value.width !== null &&
+		value.width !== undefined &&
+		value.height !== null &&
+		value.height !== undefined
+			? 'custom'
+			: value.aspectRatio;
+
+	const showScaleControl =
+		( value.aspectRatio !== null && value.aspectRatio !== undefined ) ||
+		( value.width !== null &&
+			value.width !== undefined &&
+			value.height !== null &&
+			value.height !== undefined );
+
+	return (
+		<>
+			<AspectRatioTool
+				panelId={ panelId }
+				options={ aspectRatioOptions }
+				defaultValue="auto"
+				value={ aspectRatioValue }
+				onChange={ ( nextAspectRatio ) => {
+					const { aspectRatio, ...nextValue } = value;
+
+					// 'auto' is CSS default and our default, so setting the
+					// new aspect ratio isn't needed. And 'custom' doesn't
+					// require an aspect ratio either because with and height
+					// are both set overriding any CSS aspect-ratio.
+					if (
+						nextAspectRatio !== 'custom' &&
+						nextAspectRatio !== 'auto' &&
+						nextAspectRatio !== null &&
+						nextAspectRatio !== undefined
```

Comment:
Nit: the pattern of checking `null || undefined` could be extracted to an `isNil` util. Might help cut down the verbosity.

Reply:
JavaScript has an operator that does exactly that supported all the way back to the inception of the language, if only we were allowed to use it...

That being said, I do want to clean this up, but probably by settings things with a mixture of ternary and nullish coalescing operators to flatten the values into one before checking.

Reply:
>  if only we were allowed to use it...

why aren't we?

Reply:
> why aren't we?

https://github.com/WordPress/gutenberg/blob/82d99b94263ddcaa9cdd6fb615ed331f00fdbbe6/packages/eslint-plugin/configs/es5.js#L19